mycli/main.py

```diff
@@ -387,26 +380,63 @@ def connect(self, database='', user='', passwd='', host='', port='',
 
         # Connect to the database.
 
-        try:
+        def _connect():
             try:
-                sqlexecute = SQLExecute(database, user, passwd, host, port,
-                        socket, charset, local_infile, ssl)
+                self.sqlexecute = SQLExecute(database, user, passwd, host, port,
+                                             socket, charset, local_infile, ssl)
             except OperationalError as e:
                 if ('Access denied for user' in e.args[1]):
-                    passwd = click.prompt('Password', hide_input=True,
-                                          show_default=False, type=str)
-                    sqlexecute = SQLExecute(database, user, passwd, host, port,
-                            socket, charset, local_infile, ssl)
+                    new_passwd = click.prompt('Password', hide_input=True,
+                                              show_default=False, type=str)
+                    self.sqlexecute = SQLExecute(database, user, new_passwd, host, port,
+                                                 socket, charset, local_infile, ssl)
                 else:
                     raise e
+
+        try:
+            if socket is host is port is None:
+                # Try a sensible default socket first (simplifies auth)
+                # If we get a connection error, try tcp/ip localhost
+                try:
+                    socket = '/var/run/mysqld/mysqld.sock'
+                    _connect()
+                except OperationalError as e:
+                    # These are "Can't open socket" and 2x "Can't connect"
+                    if [code for code in (2001, 2002, 2003) if code == e.args[0]]:
```

Comment:
I'm curious why we have a loop here. Would this be a simpler way of doing the same thing?
```
if e.args[0] in (2001, 2002, 2003):
```

Reply:
Yes, that's better and is how I normally do it! I was pretty exhausted when I wrote it :/